### PR TITLE
feature/units

### PR DESCRIPTION
**Pull request by builtbywill**
**Unit Support**
* adds `LETTERS` parsing rule, to allow consecutive letters and spaces to be parsed as single atoms
* adds `parse_as_unit` arg to `process_sympy()` to denote when the latex should try to parse as a unit (`Quantity`).
  * if the latex fails to parse as a unit an exception is thrown
  * if False, then every letter is treated as an individual `Symbol` (same as existing logic)
* uses and extends `sympy.physics.units` to define `latex2sympy.units` logic, defining additional units for use in the parser
  * extends the `SI` UnitSysytem with `SIE`, which adds additional dimensions and units
  * adds additional SI and Binary Prefixes
  * exports easy look-up dictionaries for all supported units and prefixes in `UNIT_ALIASES` and `PREFIX_ALIASES`
  * exports unit util methods, which are used in `process_sympy()`
* updates `process_sympy()` logic to correctly handle
  * parsing as a unit, or not by splitting apart letters
  * handling the space command as multiplication
  * combining adjacent symbols into a unit, e.g. if the prefix or unit is a latex greek letter 
* refactors util methods (methods that don't rely on other class methods) out of the `LatexToSympy` class, and into separate files in `latex2sympy.utils`

### CODE REVIEW

File: sandbox/sandbox_units_suffixes.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    'Pa',
+    'pebibit',
+    'per\\: 4\\: oz\\: serving',
+    'per\\: fl\\: oz',
+    'per\\: unit',
+    'percent',
+    'pF',
+    'photons',
+    'pico',
+    'Pokeballs\\: with\\: her\\: income',
```
Comment (lily514) [inline sandbox/sandbox_units_suffixes.py:257]: oh dear, where is `pokeballs with her income` from? is this a pop culture reference I don't know or is this from a real problem?

Comment (builtbywill) [reply sandbox/sandbox_units_suffixes.py:257]: @lily514 lol it's from a real problem, i think asking how many pokeballs someone could buy?

File: tests/unit_test.py [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+    'lb.in',
+    'ft⋅lbf',
+    'ft.lbf',
+    'ft⋅lb',
+    'ft.lb',
+
+    'psia',
+
+    'PPS',  # pulses per second
+    'kN.m',
```
Comment (lily514) [inline tests/unit_test.py:483]: 1. What unit is this?
2. What are we going to do for correctness with periods? Outside of scope for the parser, but if the student doesn't have a period but the instructor did, do we consider it correct or not?

Comment (lily514) [reply tests/unit_test.py:483]: Or wait, this is the unsupported list, was the decision we aren't supporting periods? Totally cool with that, just trying to remember

File: tests/unit_test.py [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+
+convert_to_unit_incompatible_examples = [
+    # basic - different dimensions
+    ('kg', 'm'),
+    # explicit gray != sievert
+    ('Gy', 'Sv'),
+    # rpm cannot convert
+    ('rpm', 'Hz'),
+    ('rpm', 'rad/s'),
+    # cannot covert degC or degF
```
Comment (lily514) [inline tests/unit_test.py:560]: small typo covert -> convert

File: tests/unit_test.py (hunk 2 above)
Comment (lane174) [reply tests/unit_test.py:483]: @lily514 Yeah, we decided to keep it as is. So we will throw an exception

Comment (builtbywill) [reply tests/unit_test.py:483]: @lily514 this was from a Suffix in the DB

we are not allowing periods in either the instructor unit or the student unit

**Review by lily514 — APPROVED** [review]

File: src/latex2sympy/units/unit_aliases.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
+from sympy import latex
+import sympy.physics.units.definitions.unit_definitions as sympy_units
+from sympy.physics.units.quantities import Quantity, PhysicalConstant
+from sympy.physics.units.systems.mks import all_units as mks_units, units as mks_base_units
+from sympy.physics.units.systems.mksa import all_units as mksa_units, units as mksa_base_units
+from sympy.physics.units.systems.si import all_units as si_units, units as si_base_units
+from latex2sympy.units.sie import all_units as sie_units
+from latex2sympy.units.prefixes import NEW_SI_PREFIXES, SI_PREFIXES, INFORMATION_SI_PREFIXES, BIN_PREFIXES, ALL_PREFIXES, PREFIX_ALIASES, prefix_unit
```
Comment (tragessere) [inline src/latex2sympy/units/unit_aliases.py:8]: @builtbywill Just a bit of import cleanup, PREFIX_ALIASES and json below aren't used.

Comment (builtbywill) [reply src/latex2sympy/units/unit_aliases.py:8]: @tragessere yeah those are for when the debug json output section is uncommented

**Comment by tragessere** [comment]
@builtbywill I've been messing with the sandbox and looking through the spreadsheet, and it seems to be working great for me. 

I am wondering if we should be worried about the speed of it at all though. Running sandbox.py goes from ~215ms to ~650ms, and sandbox_complex_math.py goes from ~425ms to ~840ms. Do you know if this would just run when the Azure Function starts up, or on every request?

**Comment by jkdufair** [comment]
> @builtbywill I've been messing with the sandbox and looking through the spreadsheet, and it seems to be working great for me.
> 
> I am wondering if we should be worried about the speed of it at all though. Running sandbox.py goes from ~215ms to ~650ms, and sandbox_complex_math.py goes from ~425ms to ~840ms. Do you know if this would just run when the Azure Function starts up, or on every request? (the unit list building on import part)

I wonder if we should think about something like pickling that dictionary and shipping with the code somehow. If startup time to build them is high.

**Comment by jkdufair** [comment]
> > @builtbywill I've been messing with the sandbox and looking through the spreadsheet, and it seems to be working great for me.
> > I am wondering if we should be worried about the speed of it at all though. Running sandbox.py goes from ~215ms to ~650ms, and sandbox_complex_math.py goes from ~425ms to ~840ms. Do you know if this would just run when the Azure Function starts up, or on every request? (the unit list building on import part)
> 
> I wonder if we should think about something like pickling that dictionary and shipping with the code somehow. If startup time to build them is high.

I'm seeing 400-600ms on master and about 1700ms consistently on this branch. Gonna play with pickling it just to see if that would help

**Comment by jkdufair** [comment]
With pickling, `sandbox.py` is now a steady 900ms (vs. 1700ms without) and `sandbox_complex_math` is a steady 1400ms. Without pickling, `sandbox_complex_math.py` is 2200-2300ms. Probably worth doing. I can spin off a branch as a demonstration?

**Comment by jkdufair** [comment]
Ok I pushed my proposed changes to `feature/units-pickle` if you all want to test that out. I didn't make a PR yet because I wanted to see if you want to go down that route.

**Comment by builtbywill** [comment]
> Ok I pushed my proposed changes to `feature/units-pickle` if you all want to test that out. I didn't make a PR yet because I wanted to see if you want to go down that route.

@jkdufair thanks! I just checked it out. I don't see any difference in timing on my machine using the pickled file 🤷 

@tragessere what about your machine?

I know the azure functions are much less powerful than our machines, so it will probably be worth it?

**Comment by jkdufair** [comment]
> > Ok I pushed my proposed changes to `feature/units-pickle` if you all want to test that out. I didn't make a PR yet because I wanted to see if you want to go down that route.
> 
> @jkdufair thanks! I just checked it out. I don't see any difference in timing on my machine using the pickled file 🤷
> 
> @tragessere what about your machine?
> 
> I know the azure functions are much less powerful than our machines, so it will probably be worth it?

@builtbywill Did it actually generate the pickled file? And subsequent runs after the first one had no speed improvement whatsoever? What kinds of times are you seeing?

**Comment by builtbywill** [comment]
> > > Ok I pushed my proposed changes to `feature/units-pickle` if you all want to test that out. I didn't make a PR yet because I wanted to see if you want to go down that route.
> > 
> > 
> > @jkdufair thanks! I just checked it out. I don't see any difference in timing on my machine using the pickled file 🤷
> > @tragessere what about your machine?
> > I know the azure functions are much less powerful than our machines, so it will probably be worth it?
> 
> @builtbywill Did it actually generate the pickled file? And subsequent runs after the first one had no speed improvement whatsoever? What kinds of times are you seeing?

@jkdufair ope, my timer wasn't starting at the right point in the file

okay i am seeing ~620ms down to ~370ms for sandbox_complex_math

**Comment by tragessere** [comment]
Pickle branch is cutting the increase in half on both scripts, which is pretty great. 

For the complex math one **master**: ~425ms, **units**:  ~840ms , **pickle**: ~560ms

**Comment by builtbywill** [comment]
@jkdufair @tragessere do you both approve over here then too?

**Review by tragessere — APPROVED** [review]
Yup, this branch looks great too!

**Comment by jkdufair** [comment]
Yeah based on the earlier walkthrough and a general 🫡 review, I think it's good. Let's get it merged so we can start using and testing it. :thumps:

**Pull request merged**